readme.txt
Monphi CMS
________________________________________________________________________________
License - GPL
Copyright 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012 Philip Allen
This program is distributed under the terms of the GNU General Public License

This file is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This File is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with This File; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

All files in this package follow this copyright unless otherwise specified.
________________________________________________________________________________
About
The Monphi CMS is a Content Management System that was created to simplify the design and mofication time of website design and deployment.
________________________________________________________________________________
Requirements
PHP5
Apache htaccess and mod rewrite
MySQL
________________________________________________________________________________
Installation
Unpack the files and copy them to a directory on your website. Use a web browser to access the folder you have installed the application into and follow prompts on the screen.
If you recieve any errors on the installation be sure you do not have duplicate records in the database directed for use in the setup.
________________________________________________________________________________
Download resources
http://monphi.sourceforge.net/
https://github.com/the-phil/monphi



changelog.txt
20121030
	fixed the processing of get variables. in the "Process section and page input" section of the code, around line 150. Run explode on request_URI and modify php_self with the monphi URL
	fixed auth use https cookies check with initial monphi_auth_token
	Added 404 http headers to 404 error pages, in database pull and generic versions.
	Fixed users "Last login" the feature appears have been broken adding the database query preferences. Added uid to $intUID in auth.php
	Auth preferences are now pulled from database
	Fixed file permission check on setup "checkpermissions.php" monphi.conf reports correct permissions
	Changed verbiage on checkpermissions.php test writing now reads testing read permissions.
	Changed content management for modules. where titles show (modules). e.g. change NEWS DATA to NEWS DATA (module)
	Fixed content add addslashes. Variables were changed from request to post but the routine to 
		remove addslashes, if enabled, was set to requests.
	Fixed fatal error with apache_get_modules, via function_exists. If function does not exist check for HTTP_MOD_REWRITE. If that can't be found check php_sapi_name return. Tested against CGI API based server and it appears to be returning as expected.

20110707
	Changed how modules are done. They now have the ability to modify the content section of the admin interface.
	page table no longer holds "module" rows.
	Fixed log errors with pregmatch routines in tpl and form classes. Fixes error messages not displaying with block modules when the text of the block is in the page content without the block denotion characters "[]".
	Fixed an issue with the password changes. Was using single = not double == in an if statement.
	Consolidated configuration files to two locations. Database configurations reside in conf/dbc.conf.php while everything else resides in conf/monphi.conf.php.
	Added security measures for user authentication. Using tokens, IP check, referer checks, and https cookies. Options can be changed in conf/monphi.conf.php
	Cookies should be set to http_only now. (will only work with modern browsers).
	Changed the default buttons to a slightly cleaner look.
	Fixed delete to remove module_block_links and mod_... table records.
	Auto Increment is reset with deletion of files for page, and mod_... records.

20110603
	Modified the URL handling. No more section, page, article. Simply now url.
	Backend database changes for URL, error page, index page.
	Added admin bar to the top of the page. With an "Edit This Page" link.
	Changed the logo to a code monkey.
	Authentication now handled in database.
	Setup cleaned up, default pages changed.
	Modified admin interface.
	Added users to admin interface.
	Cleaned up form process class, cleared some internal error messages.
	Moved basic files : images under img/monphi, icons now note original creators. CSS now start with monphi.
	Changed default errors, cleaner look with monphi monkey icon.

20110126
	Fixed double stripslashes if magic quote gpc is set by removing the form
		processing class arrstripslashes function calls.
	Fixed a handful of problems in the classes.
		Template (tpl) class can now handle string along with the older flat file loading.
		Error returns templates and modules have now been added and implemented into the engine.
	An [error] linker must be added into the primary template file. This will
		be used for error reporting unless suppressed (currently suppression can
		only be done via a module, will be updated later)
	All admin filenames have been changed instead of add_template change to template_add.
	Added Hide from Sitemap feature in the "Page" section
	Modules are loaded differently. Instead of using the [module] section
		in the primary page the modules are no loaded into the content section
		of each page. Each module will be loaded with [modulename].
	Modules loaded in all pages no longer require to be added to any page. Instead
		the [modulename] loading is now done in the primary HTML template file.
	Modules have more fields (fileDelete, fileAdd, fileEdit, boolAltMenu, pathAltMenu) filename is
		now fileName with a capital N in name. Was going to call it fileModule....
	Module Loading into the main menu handles new fields paths and an alternate location
		for the admin file for the module based on pathAltMenu field.
	help section in admin interface.
	Modules Sitemap has been added to installation. Changed from include to module.

20100310
	Fixed https referer for admin panel
	Fixed referer authentication bug in admin panel
	Added logout to admin panel
	Added delete feature to admin panel
	Added module edit ability
	Added future functionality fields to admin panel
	Fixed admin panel focus model via javascript
	Text Area's now use javascript and use simple editor javascript features
	Admin panel dates can now be populated by clicking the calendar icon next to them
	Admin panel may have some 
	Setup now states the "admin" user for the database requires delete capabilities.
	Started module addition to the admin panel index *will most likely be redone*
	Added some icons that are licensed under the Creative Commons
		Silk icon set 1.3 - Creative Commons Attribution 2.5 License. - mjames@gmail.com
		http://www.famfamfam.com/lab/icons/silk/

admin/auth.php
<?php
################################################################################
# File Name : auth.php                                                         #
# Author(s) :                                                                  #
#   Phil Allen phil@hilands.com                                                #
# Last Edited By :                                                             #
#   phil@hilands.com                                                           #
# Version : 2011111300                                                         #
#                                                                              #
# Copyright :                                                                  #
#   Copyright 2005, 2006, 2007, 2008, 2009, 2010, 2011 Philip Allen            #
#   This program is distributed under the terms of the GNU General Public      #
#   License                                                                    #
#                                                                              #
#   This file is free software; you can redistribute it and/or modify          #
#   it under the terms of the GNU General Public License as published          #
#   by the Free Software Foundation; either version 2 of the License,          #
#   or (at your option) any later version.                                     #
#                                                                              #
#   This File is distributed in the hope that it will be useful,               #
#   but WITHOUT ANY WARRANTY; without even the implied warranty of             #
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              #
#   GNU General Public License for more details.                               #
#                                                                              #
#   You should have received a copy of the GNU General Public License          #
#   along with This File; if not, write to the Free Software                   #
#   Foundation, Inc., 51 Franklin St, Fifth Floor,                             #
#   Boston, MA  02110-1301  USA                                                #
#                                                                              #
# External Files:                                                              #
#                                                                              #
# General Information (algorithm) :                                            #
#                                                                              #
################################################################################
#echo md5('password'); // if you locked yourself enter "5f4dcc3b5aa765d61d8327deb882cf99" into the db password for the password of "password"
################################################################################
# Grab Admin Menu                                                              #
################################################################################
$strAdmin = file_get_contents('admin_menu.tpl.html');
################################################################################
# Session information                                                          #
################################################################################
if (!ini_get("cookie_httponly"))
	ini_set("session.cookie_httponly", true);
session_start();
#echo '<br /><br /><script>document.write(document.cookie)</script>';
################################################################################
# Variables                                                                    #
################################################################################
#echo '<br /><br />'.$_COOKIE['monphi_auth_token'].'<br />'.$_COOKIE['monphi_auth_token'];
$strPath = "../";
################################################################################
# Includes                                                                     #
################################################################################
// form
if (file_exists($strPath."inc/form_process.class.inc.php")) { include_once ($strPath."inc/form_process.class.inc.php"); } else { echo 'failed to open form_process.class.inc.php'; exit;}
if (file_exists($strPath."inc/arrstripslashes.inc.php")) { include_once ($strPath."inc/arrstripslashes.inc.php"); } else { echo 'failed to open arrstripslashes.inc.php'; exit;}
// database
if (file_exists($strPath."conf/dbc.conf.php")) { include_once ($strPath."conf/dbc.conf.php"); } else { echo 'failed to open dbc.conf.php'; exit;}
if (file_exists($strPath."inc/dbc.class.inc.php")) { include_once ($strPath."inc/dbc.class.inc.php"); } else { echo 'failed to open dbc.class.inc.php'; exit;}
// auth variables
#if (file_exists($strPath."conf/auth.conf.php")) { include_once ($strPath."conf/auth.conf.php"); } else { echo 'failed to open auth.conf.php'; exit;}
#if (file_exists($strPath."conf/monphi.conf.php")) { include_once ($strPath."conf/monphi.conf.php"); } else { echo 'failed to open monphi.conf.php'; exit;}
################################################################################
# Variables                                                                    #
################################################################################
// collect data for valid referers
$intEnd = strrpos($_SERVER['PHP_SELF'], "/");
$intLength = ++$intEnd;
$strPHP_SELFFolder = substr($_SERVER['PHP_SELF'], 0, $intEnd);
if (!empty($_SERVER['HTTPS']))
	$strServer = "https://";
else
	$strServer = "http://";
// form configuration array
//----------------------------------------------------------------------------//
// arrFormConf - Form Configurations                                          //
//----------------------------------------------------------------------------//
$arrFormConf = array(
	'formFile' => "auth.tpl.html",
	'validReferer' => array( // valid referers need to be added for every page that uses auth
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'index.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'content.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'content_add.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'content_delete.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'content_edit.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'modules.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'modules_load.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'module_add.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'module_edit.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'module_delete.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'user.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'user_add.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'preferences.php',
		$strServer.$_SERVER['HTTP_HOST'].$strPHP_SELFFolder.'',
	),
	'tokenTimer' => '60',
	'fieldText' => array(
		'user' => 'User Name :',
		'pass' => 'Password :',
	),
	'fieldError' => array(
		'user' => 'User Name',
		'pass' => 'Password',
	),
	'validation' => array(
		'user' => array(
			'required',
		),
		'pass' => array(
			'required',
		),
	),
	'errorWrapper' => array(
		'0' => "\n".'<span style="color:#c00;">',
		'1' => "</span>\n"
	),
	'errorMsgWrapper' => array(
		'0' => "<br /><br />\n".'<div style="background-color:#ffffd5; border:3px #c00 solid; color:#c00; padding:5px;">'."\n\t".'<div style="font-size:1.5em; font-weight:bold; padding-bottom:.5em;"><img src="../img/monphi/logo.png" style="vertical-align:middle;" /> Uh oh, there seems to be a problem!</div>',
		'1' => "</div>\n"
	),
	'errorMsg' => array(
		'referer' => 'Invalid Referer [referer]- You must use the same form provided by the host processing it',
		'required' => '<b>[field]</b> is a required field and cannot be blank.',
		'token' => '<b>Invalid security token</b> Commonly caused by reloading the page, try resubmitting the form with the form buttons',
		'tokentime' => 'Security token ran out of time. Resubmit the form to solve this problem.',
		'useragent' => '<b>Possible session hijacking</b>, form submission has been halted.'
	)
);
//----------------------------------------------------------------------------//
// End arrFormConf - Form Configurations                                      //
//----------------------------------------------------------------------------//
################################################################################
# Reference Variables                                                          #
################################################################################
$refForm = new form_process($arrFormConf);
################################################################################
#                                                                              #
################################################################################
#echo "<pre>"; print_r($_SESSION); echo "</pre><br />\n";
#echo "REQUEST : <pre>";print_r($_REQUEST);echo "</pre><br />\n";
#echo "POST : <pre>";print_r($_POST);echo "</pre><br />\n";
#echo "GET : <pre>";print_r($_GET);echo "</pre><br />\n";
// this is just a prelim check to see if it should display the auth form or not.
if (array_key_exists('auth', $_POST))
	$boolDataSent = true;
else
	$boolDataSent = false;
################################################################################
# Core Processessing                                                           #
################################################################################
if ($boolDataSent)
{
	if (get_magic_quotes_gpc()) // incase magic quotes adds slashes we strip them
		$_POST = arrstripslashes($_POST);
	// run form error checking
	$refForm->checkValidation();
	$refForm->checkReferer();
	$refForm->checkToken();
	$refForm->checkUserAgent($refForm->strRandomSeed);
	//------------------------------------------------------------------------//
	// Data Valid Enter your Processing Code HERE                             //
	//   Use a simple if statement, if failure add                            //
	//   1. $refForm->boolError = true;                                       //
	//   2. $refForm->strErrorMsg .= "<custom error message><br />\n";        //
	//   if(<check for error>)                                                //
	//   {                                                                    //
	//       $refForm->boolError = true;                                      //
	//       $refForm->strErrorMsg .= "<custom error message><br />\n";       //
	//   }                                                                    //
	//------------------------------------------------------------------------//
	// check database for auth values
	$refDBC = new dbc($arrDBCAdmin);
	$refDBC->sql_connect();
	$strQuery = '
	SELECT
		uid, user, password
	FROM
		users
	WHERE
		active = "1"
		&&
		user = "'.mysql_real_escape_string($_POST['user']).'"
	';
	$resResult = $refDBC->query($strQuery);
	$intNumRows = mysql_num_rows($resResult);
	if ($intNumRows != 0)
	{
		$arrFetch = mysql_fetch_assoc($resResult);
			$intUID = $arrFetch['uid'];
		if (mysql_real_escape_string(md5($_POST['pass'])) != $arrFetch['password'])
		{
			$boolAuth = false;
		}
		else
		{
			$boolAuth = true;
		}
	}
	else
		$boolAuth = false;
	if (!$boolAuth) // if errors set error message
	{
		$refForm->boolError = true;
		$refForm->strErrorMsg .= "Invalid <b>Username</b> or <b>Password</b><br />\n";
		$_SESSION['auth'] = false;
	}
	//------------------------------------------------------------------------//
	// End Data Valid Enter your Processing Code HERE                         //
	//------------------------------------------------------------------------//
	if ($refForm->returnErrors()) // if error return form
	{
		$refForm->processInputData();
		echo $refForm->strTemplate;
		exit;
	}
	else // no errors generate session data
	{
		//--------------------------------------------------------------------//
		// Final Code Processing Here                                         //
		//--------------------------------------------------------------------//
		if ($boolAuth)
		{
			// get variables from database here
			$refDBC = new dbc($arrDBCAdmin);
			$refDBC->sql_connect();
//, auth_bool_use_token, auth_bool_use_HTTPS_cookies, auth_bool_check_IP_address, auth_bool_check_user_agent

			$strQuery = '
SELECT
	auth_random_seed, auth_keep_alive, auth_bool_use_HTTPS_cookies
FROM
	preferences
';
			$resResult = $refDBC->query($strQuery);
			$intNumRows = mysql_num_rows($resResult);
			if ($intNumRows != 0)
			{
				$arrFetch = mysql_fetch_assoc($resResult);
				// set all of the things that would be in monphi.conf here.
				$strRandomSeed = $arrFetch['auth_random_seed'];
				$intKeepAlive = $arrFetch['auth_keep_alive'];
				#$boolUseToken = $arrFetch['auth_bool_use_token'];
				$boolUseHTTPSCookies = $arrFetch['auth_bool_use_HTTPS_cookies'];
				#$boolCheckIPAddress = $arrFetch['auth_bool_check_IP_address'];
				#$boolCheckUserAgent = $arrFetch['auth_bool_check_user_agent'];
			}
			else
			{
				echo 'unable to retreive preferences from database';
				exit;
			}
			mysql_free_result($resResult);
			#$refDBC->sql_close();

			$_SESSION['monphi_auth'] = true; // primary auth flag
			#$_SESSION['monphi_auth_level'] = "administrator"; // this isn't checked yet, when rolls are made load roll in here instead of administrator.
			$_SESSION['monphi_auth_user'] = mysql_real_escape_string($_POST['user']); // why is this escaped?
			$_SESSION['monphi_auth_keepalive'] = time(); // timer for user keep alive. If no activity within designated time, auth is cancelled.
			$_SESSION['monphi_auth_ip_addr'] = $_SERVER['REMOTE_ADDR']; // this may cause issues with certain users/isp's
			$_SESSION['monphi_auth_user_agent'] = md5($_SERVER['HTTP_USER_AGENT'].$strRandomSeed); // Do we need the md5 and random seed?
			//$_SESSION['monphi_auth_random_seed'] = $strRandomSeed; // this should be the token, and regen and send through $_GET?
			$strToken = md5(uniqid(rand(), true));
			$_SESSION['monphi_auth_token'] = $strToken;
			if ($boolUseHTTPSCookies) // monphi.conf.php setting
				setcookie("monphi_auth_token", $strToken, time()+$intKeepAlive, '/', '', true, true);
			else
				setcookie("monphi_auth_token", $strToken, time()+$intKeepAlive, '/', '', false, true);



			$strDate = date("Y-m-d H:i:s"); // update datetime in last_login YYYY-MM-DD HH:mm:ss
			$strQuery = '
UPDATE users
SET last_login = "'.$strDate.'"
WHERE uid = "'.mysql_real_escape_string($intUID).'"
';
			$refDBC->query($strQuery);
		}
		//--------------------------------------------------------------------//
		// End Final Code Processing Here                                     //
		//--------------------------------------------------------------------//
	}
	$refDBC->sql_close();
}
else // The auth form wasn't used check session or send user form
{
	//------------------------------------------------------------------------//
	// session check                                                          //
	//------------------------------------------------------------------------//
	if (array_key_exists("monphi_auth", $_SESSION)) //only check for auth, then check for errors in auth.
	{
		// is auth set to true
		if (!$_SESSION['monphi_auth'])
		{
			// don't throw error, user may have clicked logout and our logout just removes the monphi_auth variable not key.
			#$refForm->boolError = true;
			#$refForm->strErrorMsg .= "Unable to Authorize<br />\n";
			echo $refForm->strTemplateOrig;
			exit;
		}
		// get variables from database here
		$refDBC = new dbc($arrDBCAdmin);
		$refDBC->sql_connect();
		$strQuery = '
SELECT
	auth_random_seed, auth_keep_alive, auth_bool_use_token, auth_bool_use_HTTPS_cookies, auth_bool_check_IP_address, auth_bool_check_user_agent
FROM
	preferences
';
		$resResult = $refDBC->query($strQuery);
		$intNumRows = mysql_num_rows($resResult);
		if ($intNumRows != 0)
		{
			$arrFetch = mysql_fetch_assoc($resResult);
			// set all of the things that would be in monphi.conf here.
			$strRandomSeed = $arrFetch['auth_random_seed'];
			$intKeepAlive = $arrFetch['auth_keep_alive'];
			$boolUseToken = $arrFetch['auth_bool_use_token'];
			$boolUseHTTPSCookies = $arrFetch['auth_bool_use_HTTPS_cookies'];
			$boolCheckIPAddress = $arrFetch['auth_bool_check_IP_address'];
			$boolCheckUserAgent = $arrFetch['auth_bool_check_user_agent'];
		}
		else
		{
			echo 'unable to retreive preferences from database';
			exit;
		}
		mysql_free_result($resResult);
		$refDBC->sql_close();

		// check that the users IP has not changed
		if ($boolCheckIPAddress) // monphi.conf.php setting
		{
			if ($_SESSION['monphi_auth_ip_addr'] != $_SERVER['REMOTE_ADDR'])
			{
				$refForm->boolError = true;
				$refForm->strErrorMsg .= "Remote IP changed<br />\n";
			}
		}
		// check that the user agent has not changed
		if ($boolCheckUserAgent) // monphi.conf.php setting
		{
			if ($_SESSION['monphi_auth_user_agent'] != md5($_SERVER['HTTP_USER_AGENT'].$strRandomSeed))
			{
				$refForm->boolError = true;
				$refForm->strErrorMsg .= "User Agent changed<br />\n";
			}
		}
		// check for valid token
		if ($boolUseToken) // monphi.conf.php setting
		{
			if ($_COOKIE['monphi_auth_token'] != $_SESSION['monphi_auth_token'])
			{
				$refForm->boolError = true;
				$refForm->strErrorMsg .= "Invalid Token or Token Timeout Reached<br />\n";
			}
		}
		// Incase user doesn't want to use token check session keep alive.
		$intInactiveTime = time() - $_SESSION['monphi_auth_keepalive'];
		if ($intInactiveTime >= $intKeepAlive)
		{
			$refForm->boolError = true;
			$refForm->strErrorMsg .= "Inactivity timer reached, session has been disabled<br />\n";
		}
/*
		// Token will return invalid if timer is out, no need to check twice
		$intCookieTime = time() - $_SESSION['monphi_auth_keepalive'];
		if ($intCookieTime >= $intKeepAlive)
		{
			$refForm->boolError = true;
			$refForm->strErrorMsg .= "Inactivity timer reached, cookie has been disabled<br />\n";
		}
*/
		if ($refForm->returnErrors())
		{
			// session should be dumped and data should be logged, even though we send this data
			// attacker could reattempt with data without token etc.
			// hmm maybe data should just be logged if attacked we don't want user constantly booted.
			// but this will tell the user if they have been session jacked.
			// process form submission data if errors for form reprint
			$refForm->processInputData();
			echo $refForm->strTemplate;
			exit;
		}
		else
		{
			$_SESSION['monphi_auth_keepalive'] = time(); // reset keep alive timer
			$strToken = md5(uniqid(rand(), true)); // regen token
			$_SESSION['monphi_auth_token'] = $strToken; // set token in session for verification.
			if ($boolUseHTTPSCookies) // monphi.conf.php setting
				setcookie("monphi_auth_token", $strToken, time()+$intKeepAlive, '/', '', true, true);
			else
				setcookie("monphi_auth_token", $strToken, time()+$intKeepAlive, '/', '', false, true);
		}
	}
	//------------------------------------------------------------------------//
	// if no data submission and no session display auth form.                //
	//------------------------------------------------------------------------//
	else 
	{
		echo $refForm->strTemplateOrig;
		exit;
	}
}
#echo '<br /><br /><pre>'; print_r($_SESSION); print_r($_REQUEST); echo '</pre>';
#echo '<br /><br /><pre>'; print_r($_COOKIE); print_r($_SESSION);echo '</pre>';
#echo $_COOKIE['monphi_auth_token'];
#echo '<br />';
#echo $_SESSION['monphi_auth_token'];
#echo "<pre>"; print_r($refForm->arrFields); echo "</pre>";
